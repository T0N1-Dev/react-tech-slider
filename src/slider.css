:root {
    --transition-slider-duration: 30s;
}

.wrapper {
    padding-top: 1rem;
    width: 90%;
    height: 100px;
    max-width: 1536px;
    margin-inline: auto;
    position: relative;
    overflow: hidden;
    mask-image: linear-gradient(
        to left,
        rgba(0, 0, 0, 0),
        rgba(0, 0, 0, 1) 20%,
        rgba(0, 0, 0, 1) 80%,
        rgba(0, 0, 0, 0)
    );
}

.technology-slider {
    color: white;
}

.item {
    border-radius: 6px;
    position: absolute;
    align-items: center;
    left: max(calc(-5rem * var(--amount-technologies)), 100%);
    transition: transform .3s;
    animation-name: scrollLeft;
    animation-duration: var(--transition-slider-duration);
    animation-timing-function: linear;
    animation-iteration-count: infinite;
}

.item:hover {
    transform: scale(1.1);
    .item1 {
        animation-play-state: paused;
    }
}

.item:has(img[alt="Express.js-icon"]),
.item:has(img[alt="jQuery-icon"]) {
    .technology-slider {
        filter: invert(1);
    }
}

.item1 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item2 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item3 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item4 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item5 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item6 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item7 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item8 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item9 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item10 {
  animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item11 {
  animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item12 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item13 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item14 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}
.item15 {
    animation-delay: calc(var(--transition-slider-duration) / var(--amount-technologies) * (var(--amount-technologies) - var(--item)) * -1);
}

@keyframes scrollLeft {
  to {
    left: -60rem;
  }
}

@media (max-width: 1024px) {
    .wrapper {
        margin-top: 3rem;
        height: 4rem;
    }

    .technology-slider {
        width: 3rem;
        height: 3rem;
    }

    .item {
        left: max(calc(-3rem * var(--amount-technologies)), 100%);
    }
}

@media (max-height: 699px) {
    .wrapper {
        margin-top: calc(5vh + 100px);
    }
}
