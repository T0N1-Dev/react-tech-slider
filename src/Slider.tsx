import './slider.css';
import React from 'react';

export interface Brand {
  id: number;
  name: string;
  img: string;
  width?: number; // rem
  height?: number; // rem
  style?: React.CSSProperties; 
  className?: string;
}

export interface SliderProps {
  brandsList: Brand[];
  borderWidth?: number;
  borderColor?: string;
  backgroundColor?: string;
  iconWidth?: number;
}

export const Slider: React.FC<SliderProps> = ({ 
  brandsList, 
  borderWidth = 1, 
  borderColor = '#7c05d8',
  backgroundColor = '#00000033',
  iconWidth
}) => {
  return (
    <div 
      className="wrapper"
      style={
        {
          '--amount-brands': brandsList.length,
          borderTop: `${borderWidth}px solid ${borderColor}`,
          borderBottom: `${borderWidth}px solid ${borderColor}`,
          backgroundColor: `${backgroundColor}`
        } as React.CSSProperties
      }
    >
      {brandsList.map(({ id, name, img, width, height, style, className }, index) => (
        <div
          key={id}
          className={ `item item${id}` }
          style={ { '--item': index + 1 } as React.CSSProperties }
        >
          <img
            src={img}
            alt={`${name}-icon`}
            className={`brand-slider ${className ?? ''}`}
            style={{
              ...(width || iconWidth
                ? { width: `${width ?? iconWidth}rem` }
                : {}),
              ...(height ? { height: `${height}rem` } : {}),
              ...style
            }}            
            loading="lazy"
          />
        </div>
      ))}
    </div>
  );
};
