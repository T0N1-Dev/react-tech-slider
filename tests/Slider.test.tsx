import { describe, test, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { Slider } from "../src/Slider";

describe('Slider tests', () => {
    
    const brands = [
        { id: 1, name: 'React', img: 'react.png' },
        { id: 2, name: 'Node.js', img: 'node.png' }
    ];

    test('match snapshot', () => {
        const { container } = render(< Slider brandsList={brands} />);
        expect(container).toMatchSnapshot();
    });

    test('renders all brands', () => { 
        const { getAllByRole } = render(<Slider brandsList={brands} />);
        const images = getAllByRole('img');
        expect(images).toHaveLength(brands.length);
     });

    test('applies custom styles', () => { 
        const name = 'React';
        const { container } = render(
            <Slider 
                brandsList={[{ id: 1, name: `${name}`, img: 'react.png' }]}
                borderColor="black"
                borderWidth={3}
                backgroundColor = 'red'
                iconWidth = {10}
                iconHeight = {10}
            />
        );

        const wrapper = container.querySelector('.wrapper') as HTMLElement;
        const image = screen.getByAltText(`${name}-icon`);
        
        expect(wrapper.style.borderTop).toBe('3px solid black');
        expect(wrapper.style.borderBottom).toBe('3px solid black');
        expect(wrapper.style.backgroundColor).toBe('red');
        expect(image.style.width).toBe('10rem');
        expect(image.style.height).toBe('10rem');
    });
});
