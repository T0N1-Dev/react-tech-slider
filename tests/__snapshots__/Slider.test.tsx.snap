// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slider tests > match snapshot 1`] = `
<div>
  <div
    class="wrapper"
    style="--amount-brands: 2; border-top: 1px solid #7c05d8; border-bottom: 1px solid #7c05d8; background-color: rgba(0, 0, 0, 0.2);"
  >
    <div
      class="item item1"
      style="--item: 1;"
    >
      <img
        alt="React-icon"
        class="brand-slider"
        height="50"
        loading="lazy"
        src="react.png"
        style="width: 5rem; height: 5rem;"
        width="50"
      />
    </div>
    <div
      class="item item2"
      style="--item: 2;"
    >
      <img
        alt="Node.js-icon"
        class="brand-slider"
        height="50"
        loading="lazy"
        src="node.png"
        style="width: 5rem; height: 5rem;"
        width="50"
      />
    </div>
  </div>
</div>
`;
